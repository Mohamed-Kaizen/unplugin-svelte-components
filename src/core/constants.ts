export const DISABLE_COMMENT = "<!-- unplugin-svelte-components disabled -->"
