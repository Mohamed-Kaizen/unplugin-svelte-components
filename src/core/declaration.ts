import { notNullish, slash } from "@antfu/utils"
import { promises as fs } from "fs"
import { dirname, isAbsolute, relative } from "path"

import { getTransformedPath, resolveExternalImports } from "./utils"

import type { Context } from "./context"

export async function generateDeclaration(
	ctx: Context,
	filepath: string,
	removeUnused = false
): Promise<void> {
	const items = [
		...Object.values({
			...ctx.componentNameMap,
			...ctx.componentCustomMap,
		}),
		...resolveExternalImports(ctx.options.external),
	]
	const imports: Record<string, string> = Object.fromEntries(
		items
			.map(({ from: path, as: name, name: importName }) => {
				if (!name) return undefined
				path = getTransformedPath(path, ctx)
				const related: string = isAbsolute(path)
					? `./${relative(dirname(filepath), path)}`
					: path

				let entry = `typeof import("${slash(related)}")`
				if (importName) entry += `["${importName}"]`
				else entry += `["default"]`
				return [name, entry]
			})
			.filter(notNullish)
	)

	if (!Object.keys(imports).length) return

	const lines = Object.entries({
		...imports,
	})
		.sort((a, b) => a[0].localeCompare(b[0]))
		.filter(([name]) =>
			removeUnused ? items.find((i) => i.as === name) : true
		)
		.map(([name, v]) => {
			if (!/^\w+$/.test(name)) name = `"${name}"`
			return `const ${name}: ${v}`
		})
	const code = `// generated by unplugin-svelte-components
// We suggest you to commit this file into source control
declare global {
    ${lines.join("\n    ")}
}

export {}
`
	fs.writeFile(filepath, code, "utf-8")
}
