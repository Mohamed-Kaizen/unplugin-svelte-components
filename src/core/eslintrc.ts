import type { Context } from "./context"
import { promises as fs } from "fs"
import { notNullish, resolveExternalImports } from "./utils"

export function generateESLintConfigs(ctx: Context) {
	const items = [
		...Object.values({
			...ctx.componentNameMap,
			...ctx.componentCustomMap,
		}),
		...resolveExternalImports(ctx.options.external),
	]

	const imports: Record<string, string> = Object.fromEntries(
		items
			.map(({ as: name }) => {
				if (!name) return undefined
				return [name, ctx.options.eslintrc.globalsPropValue]
			})
			.filter(notNullish)
	)

	const data = JSON.stringify({ globals: imports }, null, 4)

	fs.writeFile(
		ctx.options.eslintrc.filepath ?? "./.eslintrc-components.json",
		data,
		"utf-8"
	)
}
