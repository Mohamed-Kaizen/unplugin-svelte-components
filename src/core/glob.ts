import fg from "fast-glob"

import type { Context } from "./context"

export function searchComponents(ctx: Context) {
	const root = ctx.root

	const files = fg.sync(ctx.options.globs, {
		ignore: ["node_modules"],
		onlyFiles: true,
		cwd: root,
		absolute: true,
	})

	if (!files.length && !ctx.options.resolvers?.length)
		// eslint-disable-next-line no-console
		console.warn("[unplugin-svelte-components] no components found")

	ctx.addComponents(files)
}
