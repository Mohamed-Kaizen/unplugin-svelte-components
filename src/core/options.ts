import { slash, toArray } from "@antfu/utils"
import { join, resolve } from "path"
import { isPackageExists } from "local-pkg"

import type {
	Options,
	ResolvedOptions,
	ComponentResolver,
	ComponentResolverObject,
} from "../types"

export const defaultOptions: Omit<
	Required<Options>,
	"include" | "exclude" | "globs" | "external"
> = {
	dirs: "src/components",

	extensions: "svelte",

	deep: true,

	dts: isPackageExists("typescript"),

	directoryAsNamespace: false,

	globalNamespaces: [],

	importPathTransform: (v) => v,

	allowOverrides: false,

	collapseSamePrefixes: false,
}

function normalizeResolvers(
	resolvers: (ComponentResolver | ComponentResolver[])[]
): ComponentResolverObject[] {
	return toArray(resolvers)
		.flat()
		.map((r) =>
			typeof r === "function" ? { resolve: r, type: "component" } : r
		)
}

export function resolveOptions(
	options: Options,
	root: string
): ResolvedOptions {
	const resolved = Object.assign(
		{},
		defaultOptions,
		options
	) as ResolvedOptions
	resolved.resolvers = normalizeResolvers(resolved.resolvers)
	resolved.extensions = toArray(resolved.extensions)

	if (resolved.globs) {
		resolved.globs = toArray(resolved.globs).map((glob: string) =>
			slash(resolve(root, glob))
		)
		resolved.resolvedDirs = []
	} else {
		const extsGlob =
			resolved.extensions.length === 1
				? resolved.extensions
				: `{${resolved.extensions.join(",")}}`

		resolved.dirs = toArray(resolved.dirs)
		resolved.resolvedDirs = resolved.dirs.map((i) =>
			slash(resolve(root, i))
		)

		resolved.globs = resolved.resolvedDirs.map((i) =>
			resolved.deep
				? slash(join(i, `**/*.${extsGlob}`))
				: slash(join(i, `*.${extsGlob}`))
		)

		if (!resolved.extensions.length)
			throw new Error(
				"[unplugin-svelte-components] `extensions` option is required to search for components"
			)
	}

	resolved.dts = !resolved.dts
		? false
		: resolve(
				root,
				typeof resolved.dts === "string"
					? resolved.dts
					: "components.d.ts"
		  )

	resolved.external = resolved.external || []
	resolved.root = root

	return resolved
}
