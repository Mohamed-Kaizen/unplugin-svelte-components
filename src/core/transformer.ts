import { walk } from "estree-walker"
import MagicString from "magic-string"
import { parse } from "svelte/compiler"

import { stringifyComponentImport } from "./utils"

import type { Ast } from "svelte/types/compiler/interfaces"
import type { TransformResult } from "unplugin"
import type { Context } from "./context"
import type { Transformer } from "../types"

export default function transformer(ctx: Context): Transformer {
	return async (code, id, path) => {
		ctx.searchGlob()

		const sfcPath = ctx.normalizePath(path)

		const s = new MagicString(code)

		await transformComponent(code, s, ctx, sfcPath)

		const result: TransformResult = { code: s.toString() }

		if (ctx.sourcemap)
			result.map = s.generateMap({ source: id, includeContent: true })
		return result
	}
}

function walkAST(ast: Ast) {
	const components = new Set()
	if (ast.html && ast.html.children) {
		// eslint-disable-next-line @typescript-eslint/ban-ts-comment
		// @ts-ignore
		walk(ast.html.children, {
			// eslint-disable-next-line @typescript-eslint/no-explicit-any
			enter(node: any) {
				if (
					node.type == "InlineComponent" &&
					!/^svelte:/.test(node.name)
				) {
					components.add(node.name)
				}
			},
		})
	}
	return components
}

const resolveSvelte = (ast: Ast) => {
	const results: string[] = []

	const components = walkAST(ast)

	for (const match of components) {
		results.push(`${match}`)
	}

	return results
}

async function transformComponent(
	code: string,
	s: MagicString,
	ctx: Context,
	sfcPath: string
) {
	const ast = parse(code)

	const results = resolveSvelte(ast)

	let imports = []

	for (const name of results) {
		ctx.updateUsageMap(sfcPath, [name])

		const component = await ctx.findComponent(name, [sfcPath])

		if (component) {
			imports.push(
				stringifyComponentImport({ ...component, as: name }, ctx)
			)
		}
	}

	if (ast.instance) {
		const start = ast.instance.start

		const index = start + (code.slice(start).indexOf(">") + 1)

		const oc = s.toString()

		imports = imports.filter((item) => {
			const exists =
				oc.includes(item) || oc.includes(item.replace(/'/g, '"'))
			if (exists) return

			return item
		})

		s.overwrite(index, index + 1, `\n ${imports.join("\n")} \n`)
	} else {
		const script = `<script>${imports.join("\n")}</script>`

		s.appendLeft(ast.html.start, script)
	}
}
