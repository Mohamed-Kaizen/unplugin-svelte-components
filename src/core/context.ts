import { createFilter } from "@rollup/pluginutils"
import { slash, throttle, toArray } from "@antfu/utils"
import type fs from "fs"
import { relative } from "path"

import { generateDeclaration } from "./declaration"
import { searchComponents } from "./glob"
import {
	getNameFromFilePath,
	matchGlobs,
	parseId,
	pascalCase,
	resolveAlias,
} from "./utils"
import { resolveOptions } from "./options"
import transformer from "./transformer"

import type { UpdatePayload, ViteDevServer } from "vite"
import type {
	ImportInfo,
	Options,
	ResolvedOptions,
	Transformer,
} from "../types"
import { PreprocessorGroup } from "svelte/types/compiler/preprocess"

export class Context {
	options: ResolvedOptions
	transformer: Transformer = undefined!

	private _componentPaths = new Set<string>()
	private _componentNameMap: Record<string, ImportInfo> = {}
	private _componentUsageMap: Record<string, Set<string>> = {}
	private _componentCustomMap: Record<string, ImportInfo> = {}
	private _server: ViteDevServer | undefined

	root = process.cwd()

	sourcemap: string | boolean = true

	preprocess: PreprocessorGroup | [] = []

	constructor(private rawOptions: Options) {
		this.options = resolveOptions(rawOptions, this.root)
		this.generateDeclaration = throttle(
			500,
			false,
			this.generateDeclaration.bind(this)
		)
		this.transformer = transformer(this)
	}

	setRoot(root: string) {
		if (this.root === root) return
		this.root = root
		this.options = resolveOptions(this.rawOptions, this.root)
	}

	transform(code: string, id: string) {
		const { path, query } = parseId(id)
		return this.transformer(code, id, path, query)
	}

	setupViteServer(server: ViteDevServer) {
		if (this._server === server) return

		this._server = server
		this.setupWatcher(server.watcher)
	}

	setupWatcher(watcher: fs.FSWatcher) {
		const { globs } = this.options

		watcher.on("unlink", (path) => {
			if (!matchGlobs(path, globs)) return

			path = slash(path)
			this.removeComponents(path)
			this.onUpdate(path)
		})
		watcher.on("add", (path) => {
			if (!matchGlobs(path, globs)) return

			path = slash(path)
			this.addComponents(path)
			this.onUpdate(path)
		})
	}

	/**
	 * Record the usage of components
	 * @param path
	 * @param paths paths of used components
	 */
	updateUsageMap(path: string, paths: string[]) {
		if (!this._componentUsageMap[path])
			this._componentUsageMap[path] = new Set()

		paths.forEach((p) => {
			this._componentUsageMap[path].add(p)
		})
	}

	addComponents(paths: string | string[]) {
		const size = this._componentPaths.size

		toArray(paths).forEach((p) => this._componentPaths.add(p))
		if (this._componentPaths.size !== size) {
			this.updateComponentNameMap()
			return true
		}
		return false
	}

	addCustomComponents(info: ImportInfo) {
		if (info.as) this._componentCustomMap[info.as] = info
	}

	removeComponents(paths: string | string[]) {
		const size = this._componentPaths.size
		toArray(paths).forEach((p) => this._componentPaths.delete(p))
		if (this._componentPaths.size !== size) {
			this.updateComponentNameMap()
			return true
		}
		return false
	}

	onUpdate(path: string) {
		this.generateDeclaration()

		if (!this._server) return

		const payload: UpdatePayload = {
			type: "update",
			updates: [],
		}
		const timestamp = +new Date()
		const name = pascalCase(getNameFromFilePath(path, this.options))

		Object.entries(this._componentUsageMap).forEach(([key, values]) => {
			if (values.has(name)) {
				const r = `/${slash(relative(this.root, key))}`
				payload.updates.push({
					acceptedPath: r,
					path: r,
					timestamp,
					type: "js-update",
				})
			}
		})

		if (payload.updates.length) this._server.ws.send(payload)
	}

	private updateComponentNameMap() {
		const filter = createFilter(
			this.options.include || [/\.svelte$/],
			this.options.exclude || [
				/[\\/]node_modules[\\/]/,
				/[\\/]\.git[\\/]/,
				/[\\/]\.svelte-kit[\\/]/,
			]
		)
		this._componentNameMap = {}

		Array.from(this._componentPaths).forEach((path) => {
			if (!filter(path)) return
			const name = pascalCase(getNameFromFilePath(path, this.options))

			if (this._componentNameMap[name] && !this.options.allowOverrides) {
				// eslint-disable-next-line no-console
				console.warn(
					`[unplugin-svelte-components] component "${name}"(${path}) has naming conflicts with other components, ignored.`
				)
				return
			}

			this._componentNameMap[name] = {
				as: name,
				from: path,
				defaultImport: true,
			}
		})
	}

	async findComponent(
		name: string,
		excludePaths: string[] = []
	): Promise<ImportInfo | undefined> {
		// resolve from fs
		const info = this._componentNameMap[name]
		if (
			info &&
			!excludePaths.includes(info.from) &&
			!excludePaths.includes(info.from.slice(1))
		)
			return info
		const external = this.options.external

		if (external.length > 0) {
			const module = external
				?.filter((cp) =>
					cp?.names?.filter((cp_name) => {
						if (cp_name?.includes("as")) {
							return cp_name?.split("as")[1].trim() === name
						}
						return cp_name === name
					}).length > 0
				)
				?.at(0)
			if (!module) return undefined
			const cp_name = module?.names
				?.filter((cp_name) => {
					if (cp_name?.includes("as")) {
						return cp_name?.split("as")[1].trim() === name
					}
					return cp_name === name
				})
				?.at(0)

			if (!cp_name) return undefined

			if (cp_name?.includes("as")) {
				const [, name, alias] = cp_name.match(/^(.*) as (.*)$/) || []

				return {
					as: alias,
					from: module.from,
					name,
					defaultImport: false,
				}
			}
			return {
				as: cp_name,
				from: module.from,
				defaultImport: module.defaultImport,
			}
		}

		return undefined
	}

	normalizePath(path: string) {
		return resolveAlias(
			path,
			// eslint-disable-next-line @typescript-eslint/ban-ts-comment
			// @ts-ignore
			this.viteConfig?.resolve?.alias || this.viteConfig?.alias || []
		)
	}

	_searched = false

	/**
	 * This search for components in with the given options.
	 * Will be called multiple times to ensure file loaded,
	 * should normally run only once.
	 *
	 * @param ctx
	 * @param force
	 */
	searchGlob() {
		if (this._searched) return

		searchComponents(this)

		this._searched = true
	}

	generateDeclaration() {
		if (!this.options.dts) return
		generateDeclaration(this, this.options.dts, !this._server)
	}

	get componentNameMap() {
		return this._componentNameMap
	}

	get componentCustomMap() {
		return this._componentCustomMap
	}
}
