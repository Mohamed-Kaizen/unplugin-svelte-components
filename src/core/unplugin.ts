import { createFilter } from "@rollup/pluginutils"
import chokidar from "chokidar"
import path from "path"
import { createUnplugin } from "unplugin"
import { pathToFileURL } from "url"

import { Context } from "./context"
import { shouldTransform } from "./utils"

import type { PreprocessorGroup } from "svelte/types/compiler/preprocess"
import type { ResolvedConfig, ViteDevServer } from "vite"
import type { Options } from "../types"

export default createUnplugin<Options>((options = {}) => {
	const filter = createFilter(
		options.include || [/\.svelte$/],
		options.exclude || [
			/[\\/]node_modules[\\/]/,
			/[\\/]\.git[\\/]/,
			/[\\/]\.svelte-kit[\\/]/,
		]
	)

	const ctx: Context = new Context(options)

	return {
		name: "unplugin-svelte-components",
		enforce: "pre",

		transformInclude(id: string) {
			return filter(id)
		},

		async transform(code: string, id: string) {
			if (!code) return null
			if (!shouldTransform(code)) return null
			const result = await ctx.transform(code, id)
			ctx.generateDeclaration()
			ctx.generateESLintConfigs()
			return result
		},
		vite: {
			async configResolved(config: ResolvedConfig) {
				const configFile = path.join(config.root, "./svelte.config.js")

				const pkg =
					process.platform === "win32"
						? await import(pathToFileURL(configFile).toString())
						: await import(configFile)

				const preprocess: PreprocessorGroup | [] =
					pkg.default.preprocess || []

				ctx.preprocess = preprocess

				ctx.setRoot(config.root)

				ctx.sourcemap = true

				if (options.dts) {
					ctx.searchGlob()
					ctx.generateDeclaration()
				}
				if (options.eslintrc?.enabled) {
					ctx.generateESLintConfigs()
				}

				if (config.build.watch && config.command === "build")
					ctx.setupWatcher(chokidar.watch(ctx.options.globs))
			},
			configureServer(server: ViteDevServer) {
				ctx.setupViteServer(server)
			},
		},
	}
})
