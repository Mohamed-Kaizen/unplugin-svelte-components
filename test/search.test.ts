import { relative, resolve } from "pathe"
import { describe, expect, it } from "vitest"
import { Context } from "../src/core/context"

const root = resolve(__dirname, "../playground/sveltekit")

function cleanup(data: any) {
	return Object.values(data)
		.map((e: any) => {
			delete e.absolute
			e.from = relative(root, e.from).replace(/\\/g, "/")
			return e
		})
		.sort((a, b) => (a.as as string).localeCompare(b.as))
}

describe("search", () => {
	it("should work", async () => {
		const ctx = new Context({})
		ctx.setRoot(root)
		ctx.searchGlob()
		expect(cleanup(ctx.componentNameMap)).toMatchSnapshot()
	})
	it("should with namespace", async () => {
		const ctx = new Context({
			directoryAsNamespace: true,
			globalNamespaces: ["global"],
		})
		ctx.setRoot(root)
		ctx.searchGlob()
		expect(cleanup(ctx.componentNameMap)).toMatchSnapshot()
	})
	it("should with namespace & collapse", async () => {
		const ctx = new Context({
			directoryAsNamespace: true,
			collapseSamePrefixes: true,
			globalNamespaces: ["global"],
		})
		ctx.setRoot(root)
		ctx.searchGlob()

		expect(cleanup(ctx.componentNameMap)).toMatchSnapshot()
	})
})
