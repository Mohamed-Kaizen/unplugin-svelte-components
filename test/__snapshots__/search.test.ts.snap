// Vitest Snapshot v1

exports[`search > should with namespace & collapse 1`] = `
[
  {
    "as": "NestCounter",
    "defaultImport": true,
    "from": "src/components/nest/Counter.svelte",
  },
  {
    "as": "NestHeader",
    "defaultImport": true,
    "from": "src/components/nest/Header.svelte",
  },
]
`;

exports[`search > should with namespace 1`] = `
[
  {
    "as": "NestCounter",
    "defaultImport": true,
    "from": "src/components/nest/Counter.svelte",
  },
  {
    "as": "NestHeader",
    "defaultImport": true,
    "from": "src/components/nest/Header.svelte",
  },
]
`;

exports[`search > should work 1`] = `
[
  {
    "as": "Counter",
    "defaultImport": true,
    "from": "src/components/nest/Counter.svelte",
  },
  {
    "as": "Header",
    "defaultImport": true,
    "from": "src/components/nest/Header.svelte",
  },
]
`;
